### PR TITLE
Add Hyper-V option in windows installer

### PR DESCRIPTION
**Pull request by l0rd**
image

Here is a list of the changes included with this PR:

- `podman.msi` GUI has a radio button to select WSL or Hyper-V
- `podman.msi` GUI checkbox (to install WSL) installs Hyper-V if that's the selected provider
- `podman.msi` GUI radio buttons and checkbox are hidden if one of the following conditions are met:
  - The file `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` exists (no matter the content)
  - A pre-existing `podman.exe` file is found in `%PROGRAMFILES%/RedHat/Podman`
  - The user sets the property `SkipConfigFileCreation=1` via command line
- Fresh installation creates file `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf`:
    ```ini
    [machine]
    provider="hyperv" # or "wsl"
    ```
- Creation of the file `99-podman-machine-provider.conf` is skipped if one of those conditions are met:
  - The file already exists (no matter the content)
  - A pre-existing `podman.exe` file is found in `%PROGRAMFILES%/RedHat/Podman`
  - The user sets the property `SkipConfigFileCreation=1` via command line
- Cirrus CI `win_installer_task` has been updated to test installation with `hyperv` and uninstallation too.
- The background color of the `podman-dialog.png` has been slightly changed to match the color of the MSI package checkbox and radio-button.
- `podman-setup.exe` supports 3 new variables:
  - `MachineProvider` (valid values are `wsl` and `hyperv`)
  - `HyperVCheckbox` (valid values are `0` and `1`)
  - `SkipConfigFileCreation` (valid values are `0` and `1`)

Note that `podman-setup.exe` GUI doesn't allow to choose the provider yet. See the screenshot below and https://github.com/containers/podman/issues/22492
 
image

#### How to test it

```pwsh
# Build the installer 
cd .\contrib\win-installer\
.\build.ps1 5.0.2 dev

# Run the msi and use the GUI
msiexec /package podman.msi /l*v inst.log

# Or run the setup.exe installer selecting WSL
`.\podman-5.0.2-dev-setup.exe /install /quiet WSLCheckbox=1 /log install-wsl.log`

# Or run the setup.exe installer selecting Hyper-V
.\podman-5.0.2-dev-setup.exe /install /quiet MachineProvider=hyperv HyperVCheckbox=0 /log install-hyperv.log
```

#### Does this PR introduce a user-facing change?

```release-note
Windows installer `podman-setup.exe` supports 2 new command line variables:
- `MachineProvider`: choose the windows machine provider (can be `wsl` or `hyperv`, the default is `wsl`).
- `HyperVCheckbox`: if set to `1` the installation installs Hyper-V (the default is `0`)
- `SkipConfigFileCreation`: if set to `1` the installation skips the creation of  `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` (the default is `0`)
```

### CODE REVIEW

**Comment by baude** [comment]
LGTM but I dont really have a strong understanding for most of this to accurately review.  you need to rebase to pick up the BUD fixes from main.

**Comment by l0rd** [comment]
> LGTM but I dont really have a strong understanding for most of this to accurately review. you need to rebase to pick up the BUD fixes from main.

Thanks. Rebased.

**Review by Luap99 — COMMENTED** [review]
Should we fix https://github.com/containers/podman/issues/22411 before 5.1? If we change the location later we might have to consider backwards compatibility problems

Also should there be a third option, not create the config file at all? So that the user can use whatever the default is without having the config file. I think that is better on updates so a users does not accidental overwrite the previous default and then wonders where the machine went.

**Comment by l0rd** [comment]
> Should we fix #22411 before 5.1? If we change the location later we might have to consider backwards compatibility problems

Absolutely. If we fix #22411 before 5.1 things will be easier to handle. I am working on that.

> Also should there be a third option, not create the config file at all? So that the user can use whatever the default is without having the config file. I think that is better on updates so a users does not accidental overwrite the previous default and then wonders where the machine went.

Adding a third option (`wsl`/`hyperv`/`none`) is clever because it's a simple solution. But if we keep `wsl` as the radio-button selected by default, most users will go with that, no matter if they are installing for the first time or if they are doing an update. So that won't be effective in my opinion. We should introduce a custom action, executed before the GUI is rendered, that looks for an existing machine provider config. If it finds one, the options in the GUI will be hidden. That's for a distinct PR though.

**Comment by Luap99** [comment]
> Adding a third option (wsl/hyperv/none) is clever because it's a simple solution. But if we keep wsl as the radio-button selected by default, most users will go with that, no matter if they are installing for the first time or if they are doing an update. So that won't be effective in my opinion. We should introduce a custom action, executed before the GUI is rendered, that looks for an existing machine provider config. If it finds one, the options in the GUI will be hidden. That's for a distinct PR though.

Well I mean selecting none as default seems to best option to me. We still default to WSL if nothing is set in containers.conf so it has the same effect.

**Comment by l0rd** [comment]
> Well I mean selecting none as default seems to best option to me. We still default to WSL if nothing is set in containers.conf so it has the same effect.

A `none` that is actually `wsl` looks  confusing.

I can update this PR so that when selecting `wsl`, then no config file is created. But I don't like that neither as it unnecessarily couples installer default with machine default. None of those options looks ideal to me and I would rather keep the PR as it is and work on a proper solution (dynamic options) in a distinct PR.

File: contrib/win-installer/podman-msihooks/check.c [hunk 1]
```diff
@@ -64,3 +79,52 @@ BOOL isWSLEnabled() {
 
 	return exitCode == 0;
 }
+
+BOOL isHyperVEnabled() {
+	/*
+	 * Checks if the Windows service `vmcompute` is running to
```
Comment (n1hility) [inline contrib/win-installer/podman-msihooks/check.c:85]: Looking for service names is a good idea around the dism perm issues, but vmcompute is also used by WSL, so we need another way to detect it

**Comment by n1hility** [comment]
> > Well I mean selecting none as default seems to best option to me. We still default to WSL if nothing is set in containers.conf so it has the same effect.
> 
> A `none` that is actually `wsl` looks confusing.
> 
> I can update this PR so that when selecting `wsl`, then no config file is created. But I don't like that neither as it unnecessarily couples installer default with machine default. None of those options looks ideal to me and I would rather keep the PR as it is and work on a proper solution (dynamic options) in a distinct PR.

It's usually acceptable practice to comment out the default, and only when that differs specify a value. From that perspective you could have two options a default of wsl (no config) and hyperv (config file is placed down). Then you could follow up with a custom action that just selects hyperv as the default if it's specified in the config file.

File: contrib/win-installer/podman-msihooks/check.c (hunk 1 above)
Comment (l0rd) [reply contrib/win-installer/podman-msihooks/check.c:85]: You are right, thank you. 

I have done some tests and it looks like the service `vmms` (Hyper-V Virtual Machine Management service) is not running when `wsl` is installed and is actually running when Hyper-V is installed. Would that work in your opinion?

image

image

**Comment by l0rd** [comment]
> It's usually acceptable practice to comment out the default, and only when that differs specify a value. From that perspective you could have two options a default of wsl (no config) and hyperv (config file is placed down). Then you could follow up with a custom action that just selects hyperv as the default if it's specified in the config file.

The reason I am not comfortable with this is that we are going to be in trouble if in the future we decide to make hyper-v the podman machine default (when there is not config file) and we forget to update the installer to create a config file when wsl is selected. 

Also we are discussing about a rare case: a user that run the installer with wsl and run it again selecting hyper-v without understanding that he will lose its machine. 

And in my opinion the main problem here, is not that we are allowing a config change during installation (that's fine), but that a user cannot see both wsl and hyper machines when running `machine ls`, and I believe this is going to be addressed anyway.

**Comment by l0rd** [comment]
The PR to support system-wide containers config https://github.com/containers/common/pull/1973

**Comment by l0rd** [comment]
A quick update: 

I have updated this PR to verify if HyperV is installed using the windows service `vmms` (Hyper-V Virtual Machine Management service)

I am testing some changes to create the configuration file only if those conditions are met:
   - The file `$ProgramFiles/Red Hat/Podman/podman.exe` doesn't exist
   - The configuration file `$ProgramData/containers/containers.conf.d/99-podman-machine-provider.conf` doesn't exist

**Comment by l0rd** [comment]
The PR is updated but we should wait for https://github.com/containers/common/pull/1973 to get vendored in podman in order to merge it.

@Luap99 now the creation of `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` is skipped if one of these conditions are met:
- `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` exists (no matter the content)
- A pre-existing `podman.exe` file is found in `%PROGRAMFILES%/RedHat/Podman`
- The user sets the property `SkipConfigFileCreation=1` while running setup.exe via the command line

Also if one of the conditions above is met the provider choice is not presented to the user when running the `podman.msi` GUI:
image

@n1hility the criteria to check for Hyper-V installation is now to look at the service `vmms`. And the changes to the `podman.wxs` are now more important to support the dynamic options on the GUI. To validate that I have tested all the permutations of the variables that I think are affecting the installation:

- selected provider: `wsl/hyperv`
- install the provider: `yes/no`
- wsl is pre-installed: `yes/no`
- hyperv is pre-installed: `yes/no`
- is there a previous podman installation: `yes/no`

Also all the logic is now in the `podman.msi` GUI and not in the `setup.exe` GUI. In the short term [we should update the `setup.exe` GUI too](https://github.com/containers/podman/issues/22492) but the `msi` GUI in my opinion should be simplified (e.g. no radio buttons / checkboxes but some debugging text showing the msi properties values) as that GUI is not exposed to the users.

**Comment by Luap99** [comment]
> @Luap99 now the creation of `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` is skipped if one of these conditions are met:
> 
>     * `%PROGRAMDATA%\containers\containers.conf.d\99-podman-machine-provider.conf` exists (no matter the content)
> 
>     * A pre-existing `podman.exe` file is found in `%PROGRAMFILES%/RedHat/Podman`
> 
>     * The user sets the property `SkipConfigFileCreation=1` while running setup.exe via the command line

Sounds good to me.

**Comment by l0rd** [comment]
Changes to `containers/common` required by this PR are now vendored in `containers/podman`. So it's ok to merge when @n1hility approves it.

File: contrib/win-installer/podman-msihooks/check.c (hunk 1 above)
Comment (n1hility) [reply contrib/win-installer/podman-msihooks/check.c]: yes that sounds like it should work

**Comment by n1hility** [comment]
> Also all the logic is now in the `podman.msi` GUI and not in the `setup.exe` GUI. In the short term [we should update the `setup.exe` GUI too](https://github.com/containers/podman/issues/22492) and simultaneously simplify the `msi` GUI (e.g. no radio buttons / checkboxes but some debugging text showing the msi properties values) as that GUI is not exposed to the users.

That sounds ok as long as we aren't distributing the msi. The wsl installation aspect only requires it. Also a code signed msi looks like a rogue component (no podman icon).

I assume you tested manually both options just to verify no strange msi states? From a code review perspective it looks good just want to double check you verified it.

**Comment by l0rd** [comment]
> I assume you tested manually both options just to verify no strange msi states? From a code review perspective it looks good just want to double check you verified it.

You mean if I've tested the standalone msi? That was always the first thing I validated as that's fastest to build and test compared to the bundle. For every wsl/hyperv permutation I have tested the msi with GUI and the setup.exe via CLI (as this PR doesn't introduces changes on the bundle template side).

**Comment by l0rd** [comment]
> That sounds ok as long as we aren't distributing the msi. The wsl installation aspect requires it. Also a code signed msi on its own looks like a rogue component (no podman icon).

Agreed. That's a proposal (for a next PR) related to the observation that the standalone msi isn't published on GitHub release page and podman.io.

**Review by Luap99 — APPROVED** [review]
LGTM

**Pull request merged**